feat: nixbuild.net trait

renebarbosafl opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new integration for connecting to Nixbuild.net, enhancing the deployment capabilities with Nix and SSH configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

renebarbosafl committed:
feat: nixbuildnet trait

renebarbosafl committed:
fix: expected newline after multiline quote

renebarbosafl committed:
fix: expected newline after multiline quote

renebarbosafl committed:
fix: expected newline after multiline quote

renebarbosafl committed:
fix: patch

renebarbosafl committed:
adding env to patch

renebarbosafl committed:
fix: secret patching

renebarbosafl committed:
fix: configmap naming

renebarbosafl committed:
fix: configmap naming

renebarbosafl committed:
fix: patching volumes

renebarbosafl commented:
atlantis apply

renebarbosafl commented:
Tests: [tests.zip](https://github.com/input-output-hk/sc-dev-platform/files/13983107/tests.zip)

renebarbosafl committed:
doc: initializing

shlevy commented on infra/modules/kubevela-addons/definitions/nixbuildnet.cue:
```diff
@@ -0,0 +1,94 @@
+parameter: {
+    secretName: string
```
Ideally this wouldn't be needed, the secret would just be available everywhere (perhaps using the external-secrets thing)

renebarbosafl replied:
So can we assume the secret name will always be "nbn"? Do you suggest something else?
I can leave this parameter as optional with something as default value.

renebarbosafl replied:
@shlevy Implemented on https://github.com/input-output-hk/sc-dev-platform/pull/126/commits/ab43c58660c07dbbbaac20b1cbe60851a213f457

shlevy replied:
We can't assume `nbn` unless we also make sure the secret is always available.

renebarbosafl replied:
hmmm thinking...

If we leave "_nbn_" as the default secret name and it's not available in the namespace, the container will likely fail with `CreateContainerConfigError` until the necessary secret becomes available. The same behavior will happen if the Application is setting parameter.secretName as "_somethingelse_" and the secret "_somethingelse_" is not available.

As we don't have a default mechanism to create/distribute secrets yet I am inclined to keep that behavior and add some notes in the documentation about the secret dependency. Once we have a default mechanism for the secrets management, we can update the trait to perform it automatically.

What do you think about ?

shlevy replied:
Sounds good

renebarbosafl committed:
refact: parameter.secretName is now optional

renebarbosafl commented:
atlantis apply

renebarbosafl committed:
fix: adding nodeSelector to avoid pending jobs (waiting for ebs volume) (#127)

* fix: adding nodeselector to avoid pending jobs (waiting for ebs volume)

* fix: updating password

renebarbosafl committed:
chore: adding expose trait on cardano-node-preview (#128)

* chore: adding expose trait on cardano-node-preview

* disabling clusterIP service for preview node

falcucci committed:
add cardano wallet connector trait definition (#106)

* - addition of a new file `cardano-wallet-connector.cue` for the `kubevela-addons` module

* feat: add cardano wallet connector trait

- added a new 'kubectl_manifest' 'traitdefinition_cardano_wallet_connector' resource, allowing an application to connect and create a cardano wallet.
- the newly added trait applies to several workloads including deployments.apps, statefulsets.apps, daemonsets.apps, and jobs.batch.
- there are no conflicts specified with the new trait and it is marked as poddisruptive.
- the new trait schema is defined in 'cardano-wallet-connector.cue'.

* feat: implement cardano wallet connector configuration

- added a new configuration script for a cardano wallet connector.
- introduced new parameters: `network` and `configcloner`.
- defined separate environments for cardano node and config cloning.
- configured volumes and environment variables for both cardano node and config cloning.
- implemented logic in `#patchconfig` to populate volumes and environment variables based on the `configcloner` value.
- designed a patch for the specification template.
- added multiple containers such as `socat`, `cardano-wallet` with their respective configurations including volume mounts, arguments etc.
- implemented an additional logic for conditional initialization of the `node-config-cloner` container.
- the `node-config-cloner` container is initialized with a git command to clone the `cardano-configurations` from the input output hk github repository.

* feat: implement cardano wallet persistent volume configuration

- add a new configuration, `cardanowalletconfigs`, for the cardano wallet with defined volumes and volumemounts.
- update the volume mounts with the additional `cardanowalletconfigs` in different sections of the spec.
- change in the command parameters to include `--database` and `/wallet-db`.
- add outputs for `persistentvolumeclaim` for the volume defined in `cardanowalletconfigs`.

* feat: refactor kubevela add-on for cardano wallet connector

- add a new volume named "db" in the kubevela add-on definition for the cardano wallet connector
- change the requested storage quantity for the volumes in the kubevela add-on definition from `10gi` to `100ki`
- reference storage and class name from cardanowalletconfigs in the request for resource in the kubevela add-on definition
- add a persistent volume claim for the newly added volume "db" in the kubevela add-on definition with "readwriteonce" access mode, and the storage request referenced from cardanowalletconfigs.

* chore: enhance storage capacity in cardano wallet connector

- increase the storage request from `100ki` to `10gi` in the cardano wallet connector definition

* feat: implement "create-wallet" container with api call

- added a new container named "create-wallet"
- the container uses the `curlimages/curl:latest` image and has the pull policy set as "always"
- the container executes a post request with headers to accept json, content-type as json, and some provided json data
- the request url is `http://localhost:8090/v2/wallets`

* chore: refine cardano-wallet-connector storage and volume settings

- decreased the storage request from `10gi` to `100ki` in the cardano-wallet-connector configuration
- added volume and volumemount configurations for `cardanowalletconfigs`

* style: refactor cardano-wallet-connector configuration

- add a closing bracket in the `cardano-wallet-connector` patch block.
- remove an unnecessary closing bracket from the end of the file configuration of `cardano-wallet-connector.cue`.

* feat: refactor cardano node and wallet connector setup

- renamed `example.yaml` to `node-connector-example.yaml` in `doc/kubevela-definitions/cardano/` directory.
- introduced new files - `node-wallet-connector.md` and `node-wallet-example.yaml` in `doc/kubevela-definitions/cardano/` directory.
- in `node-wallet-example.yaml`, defined an `application` with name `dapps-certification` having various environment variables.
- added a new rule for `localdappscertification` role to list, get, create, watch, and delete pods and logs resources.
- linked above rule to `dapps-certification` service account via rolebinding in the namespace `dapps-certification-staging`.
- added service account `dapps-certification` in namespace `dapps-certification-staging`.
- the `application` uses the `cardano-wallet-connector`, `storage`, `k8s-update-strategy`, `scaler`, `https-route` and `service-account` traits.
- the `application` has a workflow step for deploying to `local-dapps-certification`, using the `local-dapps-certification` policy.
- port `80` is exposed and set as the service's port.
- the image pull policy is set to always and `iohk-ghcr-creds` is specified for imagepullsecrets.
- various configuration maps and environment variables have been defined for this application.
- a `

* docs: refactor cardano wallet connector in kubevela

- add a new documentation for the cardano wallet connector trait in kubevela with an example of usage.
- deprecate the `initcontainers` section from `cardano-wallet-connector.cue` file.

* refactor: cardano node's volume path configuration

- change the path in the volume of the cardano node from `#configclonerconfigs` to `#cardanonodeconfigs`

* refactor: cardano-wallet-connector for simplicity

- remove `configcloner` parameter from `kubevela-addons/definitions/cardano-wallet-connector.cue`
- simplify the composition of `volumes` and `envs` in the `patchconfig` section by removing conditions related to `configcloner`

* refactor: integrate network parameter in cardano-wallet name

- modify name parameter in cardano-wallet definition to append network parameter

* docs: refactor node wallet to cardano wallet in kubevela docs

- rename file `node-wallet-connector.md` to `cardano-wallet-connector.md` in `doc/kubevela-definitions/cardano/`
- rename file `node-wallet-example.yaml` to `cardano-wallet-example.yaml` in `doc/kubevela-definitions/cardano/`

* chore: simplify cardano wallet connector and remove dapps-certification

- significant reduction in environment properties for the cardano wallet connector
- removed imagepullsecrets, memory, and cpu specifications from the cardano wallet connector
- the scaling properties, https route, and service account for the cardano wallet connector have been removed
- local-dapps-certification policies and workflows have been eliminated
- the serviceaccount, role, and rolebinding definitions related to dapps-certification have been deleted

* -remove duplicated socat on cardano-wallet-connector.cue

- remove duplicated socat

* docs: refine kubevela documentation for cardano wallet connector

- simplify the description of the cardano wallet connector trait in the kubevela documentation
- remove the mention of `socat` container and specific volume attachment details from the document.

* refactor: simplify `cardano-wallet-example.yaml` component layout

- simplified the `dapps-certification` component in the `cardano-wallet-example.yaml` file by removing numerous environment variables; `imagepullsecrets`, `cpu` and `memory` settings have been removed
- removed `storage` and numerous other types from the components section
- removed entire `policies` and `workflow` sections
- removed service account, role, and role binding definitions at the end of the file

* refactor: kubernetes configuration for database handling

- removed persistentvolumeclaim `db` from `volumes`
- removed volume `db` from `volumemounts`
- removed `create-wallet` from `patch:spec:template:spec`
- removed persistentvolumeclaim `pvc-db` from `outputs`

* refactor: implement recreate strategy in wallet connector spec

- modify the `spec` of the `cardano-wallet-connector.cue` definition: add a `strategy` object with the `type` set to `recreate`, and wrap the `template: spec:` block under this new `spec`.
- make use of `retainkeys` patch strategy to maintain certain keys during the patching process.
- the update does not affect the output generation related to the persistent volume claim (pvc).

* feat: implement 'recreate' as default strategy in wallet connector

- define "recreate" as the default deployment strategy for the cardano wallet connector trait.

* chore: enhance storage capacity for cardano wallet connector

- increase the storage request for cardano wallet connector from `100ki` to `10gi`

* chore: mark conflicting traits in cardano connectors

- mark "cardano-wallet-connector" as conflicting with "cardano-node-connector" in traitdefinition
- mark "cardano-node-connector" as conflicting with "cardano-wallet-connector" in traitdefinition

* docs: refine cardano wallet connector deployment strategy

- clarify the rationale behind choosing the 'recreate' deployment strategy for the cardano wallet connector trait.

* docs: refine `cardano-node-connector` usage guidance

- clarified the usage of the `cardano-node-connector` trait in the disclaimer, recommending its standalone usage due to shared similarities.

* feat: implement configurable port for wallet connector

- add a "port" parameter to the cardano-wallet-connector module configuration
- include the new "port" parameter in the command line arguments passed to the wallet connector

* feat: cardano wallet connector for variable port support

- add a new required parameter `port` for the cardano wallet connector trait to define the port where the cardano wallet will listen for requests with default value as 8090.

* docs: refactor and update cardano wallet connector params

- refactor the parameter definition in the cardano wallet connector documentation from a list to a table format.
- change the port parameter in the cardano wallet connector definition from required to optional.

* feat: refine parameters and enhance test coverage

- change `port` to be a required parameter in `cardano-wallet-connector` configuration

renebarbosafl committed:
chore: creating new instances of cardano-node (#129)

* chore: creating new instances of cardano-node

* increase number of t3a.2xlarge nodes
